### PR TITLE
Ordinal encoder

### PR DESCRIPTION
**Pull request by kendian**
Request from Francisco. A simple script for ordinal encoding of categorical fields. If classes have a natural order, that can be supplied by the user; otherwise, the assignment of integers to classes is arbitrary.

### CODE REVIEW

File: ordinal-encoder/metadata.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            "type": "list",
+            "default": [],
+            "description": "[Optional] A list of field names or identifiers to encode. Non-categorical fields are ignored. If not supplied, or an empty list, all categorical columns are encoded."
+        },
+        {
+            "name": "mappings",
+            "type": "map",
+            "default": {},
+            "description": "[Optional] A map for the encoding of each field. Top-level keys are field names or identifiers. Values are maps of classes to integers. For example, {'col1': {'A': 0, 'B': 1, 'C': 2}}. If no mapping is given for a field, classes are assumed to have no true order and integers are assigned at random. An incomplete mapping for a field causes missing values in the new (ordinal) field."
+        }
```
Comment (jaor) [inline ordinal-encoder/metadata.json]: classes do have a "natural" order, at least one that we use everywhere else: their order in the summary.categories list of the field descriptor.  we should use that order to assign consecutive integers, if no other is given.

File: ordinal-encoder/script.whizzml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+             (make-map classes (range (count classes))))))
+    {"name" (str (field "name") "_ordinal")
+     "field" (get-field-flatline (field "name") mapping)}))
+
+;; Main workflow function
+(define (ordinal-encode dataset-id target-fields mappings)
+  (let (ds (fetch dataset-id)
+        fields (resource-fields ds)
+        fields
+          (reduce (lambda (m k) (assoc-in m [k "id"] k))
```
Comment (jaor) [inline ordinal-encoder/script.whizzml]: silly style nit: `(` should be aligned with `f` if in a new line.  but usually you would do:

```
fields (reduce (lambda (m k) ...)
               fields
               (keys fields))
```

File: ordinal-encoder/script.whizzml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                   (keys fields))
+        fields
+          (if (= 0 (count target-fields))
+              (values fields)
+              (filter (lambda (x) x)
+                      (map (lambda (x) (find-field fields x)) target-fields)))
+        fields (filter categorical-field? fields)
+        new-fields (map (lambda (x) (get-new-field x mappings)) fields)
+        args
+        {"origin_dataset" dataset-id
```
Comment (jaor) [inline ordinal-encoder/script.whizzml]: here alignment is right, but, as above, you'd usually put `{` next to args...

**Review by jaor — CHANGES_REQUESTED** [review]
looks pretty good to me... just added silly indentation nits and a request for not assigning indexes "at random".  oh, and the script needs to be added to the metadata.json at the repo's top level,  as well as the top level readme.md

File: ordinal-encoder/metadata.json (hunk 1 above)
Comment (kendian) [reply ordinal-encoder/metadata.json]: yes, "random" is incorrect. it's the order the classes are returned from the field-categories command. is that just hash order? it does not appear to be alphabetic, or the order in which they first appear, etc

**Review by osroca — COMMENTED** [review]
Please, if we are going to share this in the gallery, we might want to improve the description of this he script in metadata.json with a summary of the description in the readme plus a link to the example readme (description allows markdown)

**Comment by jaor** [comment]
in the summary, it's ordered by *counts*, breaking draws alphabetically.

**Comment by whizzmler** [comment]
(i meant breaking ties, sorry)

File: ordinal-encoder/metadata.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "Ordinal encoder.",
-    "description": "Given a dataset, encodes categorical fields using ordinal ecoding, which uses a single column of integers to represent field classes (levels).",
+    "description": "Given a dataset, encodes categorical fields using ordinal encoding, which uses a single column of integers to represent field classes (levels). It then creates a new dataset, with additional fields containing ordinal encodings of the categorical fields.\n\nIf classes have a known order (such as Like, Somewhat Like, Neutral, Somewhat Dislike, and Dislike), the integer mapping can be supplied; otherwise, integers are assigned to classes arbitrarily.\n\nFor more information, please see the [readme](https://github.com/whizzml/examples/tree/master/ordinal-encoder).",
```
Comment (osroca) [inline ordinal-encoder/metadata.json]: Thanks!

**Review by jaor — APPROVED** [review]

**Pull request merged**